### PR TITLE
Allow plugins to register target backends and make the CUDA HAL Target a plugin.

### PR DESCRIPTION
**Pull request by stellaraccident**
This isn't just an NFC refactor: it reworks the CUDA HAL target flags so that they are accessible from the compiler JIT API. Previously, they were just loose global command line flags, which are opaque to the APIs.

This reworks the target backend registration to:

* Be based around a class (TargetBackendRegistry) that can have multiple instances and a global (vs just loose functions/statics).
* Existing, non-plugin registrations go into the static global registry.
* Plugin registrations are merged into a session specific TargetBackendRegistry that combines the global and plugin targets.

There is a bit more plumbing to do for legacy entry-points, which will lose access to the CUDA HAL target if not retrofitted:

* iree-run-mlir: I'm hoping that the replacement can go in before this patch and obviate the surgery.
* MLIRInterop API: Not sure what to do here. I thought I could just outright eliminate this but comments show it is being used.
* iree-opt: Currently just working off of the global registry. I can fix this prior to landing with a special case.

### CODE REVIEW

**Comment by benvanik** [comment]
I think besides the iree-opt fix required and rebasing so you can drop your iree-run-mlir change this LGTM!

**Review by benvanik — CHANGES_REQUESTED** [review]
I think besides the iree-opt fix required and rebasing so you can drop your iree-run-mlir change this LGTM!

File: compiler/src/iree/compiler/Dialect/HAL/Target/TargetRegistry.cpp [hunk 1]
```diff
@@ -14,54 +14,82 @@ namespace IREE {
...
+                                                     CreateTargetBackendFn fn,
+                                                     bool registerStaticGlobal)
     : initFn(std::move(fn)) {
-  auto &registry = getMutableTargetRegistry();
-  if (registry.count(name) > 0) {
-    llvm::report_fatal_error(
-        "Attempting to overwrite an existing translation backend");
+  if (registerStaticGlobal) {
+    auto &registry = getMutableTargetRegistry();
+    if (registry.registrations.count(name) > 0) {
```
Comment (GMNGeoffrey) [inline compiler/src/iree/compiler/Dialect/HAL/Target/TargetRegistry.cpp]: `StringMap` has a contains method, FYI: https://llvm.org/doxygen/classllvm_1_1StringMap.html#a5bdab9e85b82bae5a3470d4fa0cf574c

and I think LLVM has a contains method for generic containers in its STL

**Review by GMNGeoffrey — COMMENTED** [review]
I haven't reviewed all the code in detail (and probably won't if Ben already has :slightly_smiling_face:), but just to confirm my understanding:

This creates IREE's first in-tree plugin, which interacts with IREE much the same as an out-of-tree plugin would. When configuring the IREE build, the still on-by-default CUDA target option controls whether this plugin gets loaded. When an external project incorporates IREE, it can similarly decide whether to load this.

Targets are registered via a global object that is explicitly initialized and explicitly referenced from plugins. Nothing happens magically at link time (except maybe LLVM flags?)

**Comment by stellaraccident** [comment]
> I haven't reviewed all the code in detail (and probably won't if Ben already has slightly_smiling_face), but just to confirm my understanding:
> 
> This creates IREE's first in-tree plugin, which interacts with IREE much the same as an out-of-tree plugin would. When configuring the IREE build, the still on-by-default CUDA target option controls whether this plugin gets loaded. When an external project incorporates IREE, it can similarly decide whether to load this.

There are a couple of in-tree sample plugins under /samples (and only enabled if building samples), but approximately yes. This is the first of the always-enabled in-tree plugins. We may make it more possible to disable in the future but always-on matches the current behavior...

> 
> Targets are registered via a global object that is explicitly initialized and explicitly referenced from plugins. Nothing happens magically at link time (except maybe LLVM flags?)

It does arrange to link the library in, but there is no magic: the list of available plugins is expanded to C entry points which are then just called (passing a registration object).

File: compiler/src/iree/compiler/Tools/BUILD.bazel [hunk 2]
```diff
@@ -178,20 +178,14 @@ iree_compiler_cc_library(
         "IREE_HAVE_ROCM_TARGET",
         "IREE_HAVE_VMVX_TARGET",
         "IREE_HAVE_VULKANSPIRV_TARGET",
-    ] + select({
-        ":cuda_enabled": ["IREE_HAVE_CUDA_TARGET"],
-        "//conditions:default": [],
-    }),
+    ],
```
Comment (GMNGeoffrey) [inline compiler/src/iree/compiler/Tools/BUILD.bazel:181]: It looks like this breaks Cuda in Bazel again? The Bazel macro `iree_compiler_register_plugin` is a noop: https://github.com/openxla/iree/blob/afff73aa4004ee8a9ba6a6fc434a1ba4deba3705/build_tools/bazel/build_defs.oss.bzl#L76

**Review by GMNGeoffrey — COMMENTED** [review]
> This is the first of the always-enabled in-tree plugins. We may make it more possible to disable in the future but always-on matches the current behavior...

Looks to me like it's "on by default" not "always on", no? Always on would be a problem

https://github.com/openxla/iree/blob/2c4972b3b7b8872dbe267d36a776ee0bfa5b25b9/compiler/plugins/iree_compiler_plugin_group.cmake#L7-L9

> It does arrange to link the library in, but there is no magic: the list of available plugins is expanded to C entry points which are then just called (passing a registration object).

I'm mostly trying to make sure we don't go back to `alwayslink` hell.

File: compiler/src/iree/compiler/Tools/BUILD.bazel (hunk 2 above)
Comment (stellaraccident) [reply compiler/src/iree/compiler/Tools/BUILD.bazel:181]: Yeah, that is one of the things I need to fix before this pr is done. Missed it before firing off.

**Comment by stellaraccident** [comment]
> I'm mostly trying to make sure we don't go back to `alwayslink` hell.

It's not alwayslink, just normal linking combined with a generated header that calls the registration functions.

**Comment by stellaraccident** [comment]
Update on this. We've eliminated the iree-run-mlir and MLIRInterop API divergence (eliminated both in favor of the embedding API), so this PR can move forward now. Just need to rebase and finish it.

**Comment by silvasean** [comment]
How is this different from the LLVMGPU Codegen? Shouldn't all of the `compiler/src/iree/compiler/Codegen/LLVMGPU` directory be moved into this plugin too?

**Comment by stellaraccident** [comment]
Eventually, but even just pulling the instantiation point in gets us a measure of control we didn't have before. Practically, this was the most direct way to get some of the LLVMGPU/CUDA flags exposed to the compiler library in a way that can be used by JIT compilers (previously they were just loose, global LLVM options, which are not compatible with being set from JIT/libraries).

**Comment by jpienaar** [comment]
Nice! Sorry I missed this while away, I'll check this out ~soon.

**Review by benvanik — APPROVED** [review]

**Pull request merged**